cloudsync/tasks_test.py

```diff
@@ -229,3 +236,106 @@ def test_stream_to_s3_no_video():
     """Test DoesNotExistError"""
     with pytest.raises(Video.DoesNotExist):
         stream_to_s3(999999)  # pylint: disable=no-value-for-parameter
+
+
+@pytest.mark.django_db
+@mock_s3
+@override_settings(LECTURE_CAPTURE_USER='admin')
+def test_monitor_watch(mocker, user):  # pylint: disable=unused-argument,redefined-outer-name
+    """Test the Watch bucket monitor task"""
+    mocker.patch.multiple('cloudsync.tasks.settings',
+                          ET_PRESET_IDS=('1351620000001-000061', '1351620000001-000040', '1351620000001-000020'),
+                          AWS_REGION='us-east-1', ET_PIPELINE_ID='foo')
+    mocker.patch('cloudsync.tasks.VideoTranscoder.encode')
+    mocker.patch('dj_elastictranscoder.transcoder.Session')
+    mocker.patch('celery.app.task.Task.update_state')
+    mocker.patch('cloudsync.api.process_transcode_results')
+    mocker.patch('cloudsync.api.get_et_job',
+                 return_value={'Id': '1498220566931-qtmtcu', 'Status': 'Complete'})
+    user.username = 'admin'
+    user.save()
+    s3 = boto3.resource('s3')
+    s3c = boto3.client('s3')
+    upload = 'MIT-6.046-2017-Spring-lec-mit-0000-2017apr06-0404-L01.mp4'
+    s3c.create_bucket(Bucket=settings.VIDEO_S3_WATCH_BUCKET)
+    s3c.create_bucket(Bucket=settings.VIDEO_S3_BUCKET)
+    bucket = s3.Bucket(settings.VIDEO_S3_WATCH_BUCKET)
+    bucket.upload_fileobj(io.BytesIO(os.urandom(6250000)), upload)
+    assert s3c.get_object(Bucket=bucket.name, Key=upload) is not None
+    monitor_watch_bucket.delay()
+    new_video = Video.objects.get(title=upload)
+    new_videofile = new_video.videofile_set.get(encoding=EncodingNames.ORIGINAL)
+    assert new_videofile.bucket_name == settings.VIDEO_S3_BUCKET
+    with pytest.raises(ClientError):
+        s3c.get_object(Bucket=bucket.name, Key=upload)
+
+
+@pytest.mark.django_db
+@mock_s3
+@override_settings(LECTURE_CAPTURE_USER='admin')
+def test_monitor_watch_exceptions(mocker, user):  # pylint: disable=unused-argument,redefined-outer-name
+    """
+    Test that an exception is correctly handled
+    """
+    s3 = boto3.resource('s3')
+    s3c = boto3.client('s3')
+    upload = 'MIT-6.046-2017-Spring-lec-mit-0000-2017apr06-0404-L01.mp4'
+    s3c.create_bucket(Bucket=settings.VIDEO_S3_WATCH_BUCKET)
+    bucket = s3.Bucket(settings.VIDEO_S3_WATCH_BUCKET)
+    bucket.upload_fileobj(io.BytesIO(os.urandom(6250000)), upload)
+    with pytest.raises(User.DoesNotExist) as exc:
+        monitor_watch_bucket.delay()
+        assert 'User matching query does not exist' in str(exc)
+    user.username = 'admin'
+    user.save()
+    with pytest.raises(ClientError) as exc:
+        monitor_watch_bucket.delay()
+    assert 'NoSuchBucket' in str(exc)
+    mocker.patch('cloudsync.tasks.process_watch_file',
+                 side_effect=ValueError('Fake ValueError'))
+    with pytest.raises(ValueError) as exc:
+        monitor_watch_bucket.delay()
+    assert 'Fake ValueError' in str(exc)
```

Comment:
these tests are not sufficient: we need to test each of the major variation of the APIs and especially all the cases where they can fail.
we also need to assert for each mocked function, that it has been called with the expected parameters.